File: Lessons/Python/Unit 1/lesson01.py
### CUSTOM FUNCTIONS ###
'''
So far, we have looked at the built in 'print()' function. Python has
many other built in functions, but a lot of them require more extensive
knowledge of the language, so for now we are instead finally going to
look into making our own functions.

Functions are similar to variables in that they can be named anything.
Most people use the same naming convention for functions as they do
with variables. The difference is, when you define (create) a function,
you must tell python what you are doing. To tell python you are making
a function, the syntax is as follows:
'''

def functionName():
    #do stuff
    pass

'''
Notice how just like if statements, we have a colon and white space
indicating a block of code will follow. That block of code will
replace the 'do stuff' comment. Below that comment, we see another
python keyword 'pass'. The 'pass' keyword means literally 'do nothing'
the only reason it is there is because a function cannot be empty in
python, it must contain some code.

So what is the point of a function? Remember how in the last lesson
we ended up typing the same code several times? A function's purpose
is to prevent you from needing to write the same code more than
neccessary. For example, say we wanted to print 'Hello, world!'
multiple times without typing 'print("Hello, world!")' over and over.
'''

def phw():
    print("Hello world!")

'''
We have created our function, now it is time to call it. Remember how
we called 'print()'? We must put a pair of parenthesis at the end of
the name of the function to call it.
'''

phw()

# We can do it as many times as we like!
phw()
phw()
phw()

# NOTE: A function must be defined (created) before it is
#		called, as python reads files from top to bottom
#		and won't know what to do if you call a function before
#		it is created.

'''
This is cool and all, but it does not actually do anything useful. The
true power of functions is when you have one that does some sort of
computing with data that you provide to it. Remember how we could pass
a string or a number to 'print()' to make it print something on the
terminal? You can make your custom functions take data in as well. The
information you are passing to your function are called 'parameters'.

Parameters are what go inside of your function's parenthesis. When you
create your function, you will put the parameters in the parenthesis as
variables. You can then use these parameters as if you declared them as
variables with an '=' sign. When you call the function, you can pass
anything you want as long it is the same amount of parameters and the
right order that you want, or else you might get an unwanted outcome.

Let's demonstrate this by creating a simple function that adds two
numbers.
'''

def add(x, y):
    print(x + y)

# We have created a function that takes two numbers and prints their stum
# Now, lets call it
add(2, 1)


### MORE BOOLEAN OPERATORS ###
'''
Before we continue having fun with functions, we should take some time
to learn more complex boolean operators. This will also serve as a
review, as later we will begin combining multiple aspects of what we
have learned.
'''

# Set some variables to perform boolean logic
x = 1
y = 2
z = 1

# Review what we have seen before
print(x == 1)
print(x != y)
print(x != z )

# New operators
print(y > x) # checks if y is greater than x
print(y < x) # checks if y is less than x
print(x >= z) # checks if x is greater than or equal to z
print(y <= 2) # checks if y is less than or equal to 2

'''
There are three other operators that are also very important but
will require a more complex explanation.

The first operator is the not operator. It is the most simple of
the three. The not operator comes before a boolean operation and
inverts/flips the result. For example, if we have an operation that
is 'True', placing a 'not' before it will flip it to 'False'
and vice versa.
'''
x = True
print(x)
print(not x)

y = 2
print(not(y != 2))

# A more complex example. Challenge: This can be written in
# a better way. Can you write it better?
x = 5
y = 6
if(not(x + y == 10)):
    print("The sum of 6 and 5 is not 10")

'''
The second operator is the boolean 'and'. An and can link two
booleans together to form their own boolean. In order for a boolean
'and' to be true, both sides of the operator must be true. For
example, if I asked you to bring me a keyboard and a mouse, and you
only brought me a keyboard, would you have done what I asked? No, you
must bring me a keyboard AND a mouse in order to have done what I
asked.
'''
x = True
y = True
print(x and y) # This should return True, as both x and y are true

z = False
print(x and z) # This should return False, as not both x and z are true

'''
The third and alst operator we are going over is the boolean 'or'.
Like 'and', 'or' links two boolean functions together. However, with
'or', as long as at least one side is true, then the 'or' will be
true. Using the same example as before, if I asked you to get me a
keyboard OR a mouse, I would be okay with you bringing just a keyboard,
just a mouse, or both. But I would not be okay with you bringing me
neither.
'''
x = True
y = False
z = True

print(x or y)
print(x or z)
print(y or False)

File: Lessons/C/Unit 1/lesson01.c
/*
In this lesson we will be learning about variables in C,
a new operator that isn't in python, and little bit on
arrays. This is a lot of content, but it is important
knowledge for the road ahead. Let's begin.
*/

// First we must include our io library
#include <stdio.h>

// Lets start by making our main function
int main()
{

	/*
	We will now declare two variables in two different
	ways. The first way is to do it all in one line. We
	must tell the compiler what kind of data will be
	assigned to the variable first. In our case, we will
	have an integer. Then, we can name the variable. The
	naming conventions are the same as in python. Let's
	assign the value of 5 to our variable.
	*/
	int num = 5;

	/*
	The second way to name your variable is to
	initialize it (give it a data type and name) first
	and then assign it a value later. This may not seem
	very useful, but this will become a very useful tool
	later on.
	*/
	int num2;
	// We can now assign it a value.
	num2 = 5;


	/*
	Now we can print these values. Printing in C is very
	different in C compared to python. All 'printf' does
	is print a string, not numbers. However, it can take
	numbers as an argument when formatting the string.
	That is where the 'f' in 'printf' comes in, it means
	'format'.

	The first argument is the string to be formatted.
	Notice that where our values would go, there are two
	'%d'. The % tells printf that we are inserting a
	value in that position. The following letter tells
	printf what kind of value will be going there. In
	this case, the 'i' means 'signed integer'. After we
	give it our string, we give it the two values to
	insert as seperate arguments.
	*/
	printf("num is %i and num2 is %i", num, num2);

	/*
	Now for some math. C has all of the same math
	operators as pyhton, however it has one that python
	does not have. Like python, you could add one to the
	num variables like this:
	*/
	num += 1;

	// Or you could do it like this
	num2++;
	/*
	This is the standard way of adding one to a numeric
	value in c. You could also subtract one from it with
	num2--. Let's print out what we have so far.
	*/

	// Let's print a newline first to separate this line
	// from the last.
	printf("\n");

	printf("num is %i and num2 is %i", num, num2);


	/*
	Now lets get into the concept of arrays. If you
	remember lists from python, arrays are similar in
	that they hold data. Where they differ is that
	arrays are of fixed size, they cannot change size
	without being very familiar with how C works and
	knowing some work arounds.

	Let's look at two ways to make an array. The first
	is to make the array all together on one line. This
	is done by initializing the array by telling the
	compiler we are making an array with square brackets
	and then writing out the elements of the array with
	curly brackets.
	*/
	int numbers[] = {5, 6, 3, 2};

	/*
	The other way to make an array is to tell the
	compiler that the array will be a certain size. Then
	you can assign variables to the array at their
	specific index.
	*/
	int numbers2[3];
	numbers2[0] = 3;
	numbers2[1] = 4;
	numbers2[2] = 5;

	/*
	Now lets check and make sure it is working. There is
	not an easy way to print arrays of numbers, so for
	now lets print just the first element of the arrays.
	Remember, counting starts at 0.
	*/
	printf("\n");
	printf("The first element of numbers is %i and the first element of numbers2 is %i", numbers[0], numbers2[0]);

	/*
	The reason I am showing you arrays now is because
	string do not exist in basic C. A string in C is
	actually an array of 'char' data. A 'char' is a
	character, meaning a single symbol. Let's set a char
	variable and print it to the screen.
	*/

	char letter = 'a';
	char symbol = '~';

	printf("\n");

	// Notice how we now use %c for chars
	printf("letter is %c and symbol is %c", letter, symbol);

	/*
	Now we can make a character array for us to print.
	Lets make one that says 'Hello, world!'. Notice how
	I can use quotes to list the elements of the array.
	I am only able to do this for character arrays.
	*/
	char message[] = "Hello, world!";
	printf("\n");
	printf(message);

	// Return our error message
	return 0;
}